### PR TITLE
[storage] Custom serialization / deserialization for Jellyfish

### PR DESCRIPTION
**Pull request by zamsden**
## Motivation

This adds custom serde for Jellyfish which has a more efficient encoding
of version and compact bitfield encoding for nibble hashmap keys.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/libra/libra/blob/master/CONTRIBUTING.md#pull-requests)?

Mostly.

## Test Plan

Existing CI / unit tests.

### CODE REVIEW

File: storage/jellyfish_merkle/src/node_type/mod.rs [hunk 1]
```diff
@@ -583,4 +646,47 @@ pub enum NodeDecodeError {
...
+where
+    T: Read,
+{
+    let mut num = 0u64;
+    for _ in 0..=8 {
+        let byte = reader.read_u8()?;
+        let more = (byte & 0x80) != 0;
+        num = num << 7 | (byte & 0x7f) as u64;
+        if !more {
+            break;
```
Comment (msmouse) [inline storage/jellyfish_merkle/src/node_type/mod.rs]: `Ok(())` here ...

File: storage/jellyfish_merkle/src/node_type/mod.rs [hunk 2]
```diff
@@ -583,4 +646,47 @@ pub enum NodeDecodeError {
...
+        }
+    }
+    Ok(num)
```
Comment (msmouse) [inline storage/jellyfish_merkle/src/node_type/mod.rs]: ... and `unreachable()` here?

File: storage/jellyfish_merkle/src/node_type/mod.rs (hunk 1 above)
Comment (zamsden) [reply storage/jellyfish_merkle/src/node_type/mod.rs]: I'll need to `return Ok(num)` which seems a little weird.

File: storage/jellyfish_merkle/src/node_type/mod.rs (hunk 2 above)
Comment (zamsden) [reply storage/jellyfish_merkle/src/node_type/mod.rs]: With a bad encoding, this is reachable.  Should I bother checking for and setting errors on bad encodings?

Comment (zamsden) [reply storage/jellyfish_merkle/src/node_type/mod.rs]: I decided to check for bad encodings, better practice.

**Comment by wqfish** [comment]
I thought we are still using something like [this](https://github.com/libra/libra/blob/ae096aeec5e19ed58c3c3250db13bdef354f43bd/storage/sparse_merkle/src/node_serde/mod.rs), but just realized that it was removed when we migrated to jellyfish.

File: storage/jellyfish_merkle/src/node_type/mod.rs [hunk 3]
```diff
@@ -583,4 +646,53 @@ pub enum NodeDecodeError {
...
+    for i in 0..8 {
+        let byte = reader.read_u8()?;
+        let more = (byte & 0x80) != 0;
+        num = num | (u64::from(byte & 0x7f) << i * 7);
+        if !more {
+            return Ok(num);
+        }
+    }
+    // Last byte is encoded as is.
+    let byte = reader.read_u8()?;
```
Comment (msmouse) [inline storage/jellyfish_merkle/src/node_type/mod.rs]: 👏

File: storage/jellyfish_merkle/src/node_type/node_type_test.rs [hunk 4]
```diff
@@ -82,6 +82,22 @@ fn test_encode_decode() {
     }
 }
 
+proptest! {
+    #[test]
+    fn test_u64_varint_roundtrip(input in any::<u64>()) {
```
Comment (msmouse) [inline storage/jellyfish_merkle/src/node_type/node_type_test.rs]: 👏

@bors-libra delegate+

File: storage/jellyfish_merkle/src/node_type/mod.rs [hunk 5]
```diff
@@ -316,6 +317,68 @@ impl InternalNode {
...
+                Child::new(
+                    HashValue::from_slice(&reader.get_ref()[pos..pos + size_of::<HashValue>()])?,
+                    version,
+                    (leaf_bitmap & child_bit) != 0,
+                ),
+            );
+            reader.seek(SeekFrom::Current(size_of::<HashValue>() as i64))?;
+            existence_bitmap &= !child_bit;
+        }
+        assert_eq!(existence_bitmap, 0);
```
Comment (wqfish) [inline storage/jellyfish_merkle/src/node_type/mod.rs]: Is it also good to add something like `ensure!(reader.position() == len)`?

File: storage/jellyfish_merkle/src/node_type/mod.rs [hunk 6]
```diff
@@ -583,4 +646,53 @@ pub enum NodeDecodeError {
...
+
+/// Helper function to serialize version in a more efficient encoding.
+/// We use a super simple encoding - the high bit is set if more bytes follow.
+fn serialize_u64_varint(mut num: u64, binary: &mut Vec<u8>) {
+    for _ in 0..8 {
+        let low_bits = num as u8 & 0x7f;
+        num >>= 7;
+        let more = (num > 0) as u8;
+        binary.push(low_bits | more << 7);
+        if more == 0 {
```
Comment (wqfish) [inline storage/jellyfish_merkle/src/node_type/mod.rs]: I might prefer

```
let more = num > 0;
binary.push(low_bits | (more as u8) << 7);
if !more {
```

Seems a tiny bit more symmetric given that `more` in `deserialize_u64_varint` is a boolean.

**Pull request closed without merging**